### PR TITLE
[TPU][V1] Guided decoding on TPU

### PR DESCRIPTION
**Pull request by carlesoctav**
Hi, this is just a draft PR for issue #11104.  
I'm not new to TPU and JAX, but I'm new to LLM-serving-engine and PyTorch XLA, so I have a lot of questions here.

I've created a simple implementation for guided decoding on TPU. Here are the issues I encountered:

1. One problem I noticed is that we cannot use `sample_from_hidden` because it requires modified logits (bitmasking the logits that don't follow the grammar). However, I believe `sample_from_hidden` is actually a jitted function (?), as it was captured when the `capture_model` function runs. Performance-wise, it seems to slow down significantly (I just eyeballed the tokens per second from the logger) when the request requires guided decoding compared to the default one. Should we capture the jitted function in a smaller scope, such as `compute_logits` and `sample`? I need to profile this further to identify which part of the code is the bottleneck. Is there any code that can help with profiling this?

2. The XGrammar library doesn't support bitmasking directly on TPU, so I had to move the logits to the CPU first and then return them to the TPU for sampling.

Any feedback is welcome.

Thank you.

### CODE REVIEW

**Comment by robertgshaw2-redhat** [comment]
Thanks for your contribution!

I think that:
- this could make sense as a pathway to enable the feature
- this could have very bad performance (since the bitmask + sampling happens on CPU)

It would be helpful to know how much overhead you see from this.

**Comment by robertgshaw2-redhat** [comment]
I also wonder if there is a way we could work with xgrammar to implement this on TPU

**Comment by robertgshaw2-redhat** [comment]
anyways defer to @aarnphm and @russellb for review

**Comment by russellb** [comment]
In general, the code being changed is what I would expect to see.

I think the main question is how good/bad the performance is and if more work is needed before it makes sense to turn on the feature.

**Comment by russellb** [comment]
cc @Ubospica -- it would be great if we could find a way to get fast support for TPU in xgrammar

**Comment by aarnphm** [comment]
> I also wonder if there is a way we could work with xgrammar to implement this on TPU

I think we should have the TPU kernel implemented on xgrammar

**Comment by carlesoctav** [comment]
can we write the `apply_token_bitmask_inplace_kernel` in pallas and let the torch_xla `make_kernel_from_pallas` function do the rest?

**Comment by carlesoctav** [comment]
i run some benchmark with the current implementation, and here's the results:

server: v4-8
server-command: python -m vllm.entrypoints.openai.api_server --model Qwen/Qwen2.5-7B-Instruct -tp 4  --max-model-len 8192 --max-num-batched-tokens 256 --max-num-seqs 256

client: ./run_structured_output_benchmark.sh 


------------------------------------------------

============ Serving Benchmark Result ============
Successful requests:                     1000
Benchmark duration (s):                  107.59
Total input tokens:                      273102
Total generated tokens:                  58352
Request throughput (req/s):              9.29
Output token throughput (tok/s):         542.34
Total Token throughput (tok/s):          3080.64
---------------Time to First Token----------------
Mean TTFT (ms):                          216.86
Median TTFT (ms):                        204.70
P99 TTFT (ms):                           423.94
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          171.07
Median TPOT (ms):                        128.82
P99 TPOT (ms):                           1225.24
---------------Inter-token Latency----------------
Mean ITL (ms):                           107.25
Median ITL (ms):                         116.24
P99 ITL (ms):                            146.57
==================================================
correct_rate(%) 84.2

Completed benchmark with QPS: 10


------------------------------------------------


============ Serving Benchmark Result ============
Successful requests:                     1000
Benchmark duration (s):                  86.53
Total input tokens:                      273102
Total generated tokens:                  57478
Request throughput (req/s):              11.56
Output token throughput (tok/s):         664.23
Total Token throughput (tok/s):          3820.28
---------------Time to First Token----------------
Mean TTFT (ms):                          1691.63
Median TTFT (ms):                        925.22
P99 TTFT (ms):                           5464.36
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          381.00
Median TPOT (ms):                        266.74
P99 TPOT (ms):                           4524.30
---------------Inter-token Latency----------------
Mean ITL (ms):                           227.68
Median ITL (ms):                         255.08
P99 ITL (ms):                            310.35
==================================================
correct_rate(%) 84.3

Completed benchmark with QPS: 15

----------------------------------------------------------------

============ Serving Benchmark Result ============
Successful requests:                     1000
Benchmark duration (s):                  84.01
Total input tokens:                      273102
Total generated tokens:                  57389
Request throughput (req/s):              11.90
Output token throughput (tok/s):         683.12
Total Token throughput (tok/s):          3933.95
---------------Time to First Token----------------
Mean TTFT (ms):                          7819.44
Median TTFT (ms):                        7185.06
P99 TTFT (ms):                           20134.09
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          381.37
Median TPOT (ms):                        274.79
P99 TPOT (ms):                           2212.31
---------------Inter-token Latency----------------
Mean ITL (ms):                           245.06
Median ITL (ms):                         262.64
P99 ITL (ms):                            325.29
==================================================
correct_rate(%) 84.2

Completed benchmark with QPS: 20



----------------------------------------------------------------

============ Serving Benchmark Result ============
Successful requests:                     1000
Benchmark duration (s):                  82.46
Total input tokens:                      273102
Total generated tokens:                  57249
Request throughput (req/s):              12.13
Output token throughput (tok/s):         694.27
Total Token throughput (tok/s):          4006.23
---------------Time to First Token----------------
Mean TTFT (ms):                          11634.63
Median TTFT (ms):                        11344.79
P99 TTFT (ms):                           28409.27
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          404.67
Median TPOT (ms):                        272.93
P99 TPOT (ms):                           4440.12
---------------Inter-token Latency----------------
Mean ITL (ms):                           247.11
Median ITL (ms):                         260.56
P99 ITL (ms):                            331.74
==================================================
correct_rate(%) 84.39999999999999

Completed benchmark with QPS: 25


----------------------------------------------------------------

============ Serving Benchmark Result ============
Successful requests:                     1000
Benchmark duration (s):                  85.30
Total input tokens:                      273102
Total generated tokens:                  57682
Request throughput (req/s):              11.72
Output token throughput (tok/s):         676.26
Total Token throughput (tok/s):          3878.11
---------------Time to First Token----------------
Mean TTFT (ms):                          23213.34
Median TTFT (ms):                        23266.43
P99 TTFT (ms):                           50684.18
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          421.85
Median TPOT (ms):                        280.41
P99 TPOT (ms):                           7657.69
---------------Inter-token Latency----------------
Mean ITL (ms):                           255.30
Median ITL (ms):                         259.13
P99 ITL (ms):                            340.72
==================================================
correct_rate(%) 84.39999999999999

Completed benchmark with QPS: 50


----------------------------------------------------------------

============ Serving Benchmark Result ============
Successful requests:                     1000
Benchmark duration (s):                  87.14
Total input tokens:                      273102
Total generated tokens:                  57810
Request throughput (req/s):              11.48
Output token throughput (tok/s):         663.44
Total Token throughput (tok/s):          3797.62
---------------Time to First Token----------------
Mean TTFT (ms):                          24319.86
Median TTFT (ms):                        24451.97
P99 TTFT (ms):                           54490.02
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          423.79
Median TPOT (ms):                        292.22
P99 TPOT (ms):                           4555.85
---------------Inter-token Latency----------------
Mean ITL (ms):                           264.07
Median ITL (ms):                         268.12
P99 ITL (ms):                            342.77
==================================================
correct_rate(%) 84.2

Completed benchmark with QPS: 60

----------------------------------------------------------------
============ Serving Benchmark Result ============
Successful requests:                     1000
Benchmark duration (s):                  126.67
Total input tokens:                      273102
Total generated tokens:                  57100
Request throughput (req/s):              7.89
Output token throughput (tok/s):         450.77
Total Token throughput (tok/s):          2606.74
---------------Time to First Token----------------
Mean TTFT (ms):                          49805.52
Median TTFT (ms):                        51288.06
P99 TTFT (ms):                           84605.17
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          498.67
Median TPOT (ms):                        330.97
P99 TPOT (ms):                           4561.90
---------------Inter-token Latency----------------
Mean ITL (ms):                           310.90
Median ITL (ms):                         282.89
P99 ITL (ms):                            393.64
==================================================
correct_rate(%) 84.39999999999999

Completed benchmark with QPS: 70

**Comment by aarnphm** [comment]
> can we write the `apply_token_bitmask_inplace_kernel` in pallas and let the torch_xla `make_kernel_from_pallas` function do the rest?

I think this makes sense, would be great to contribute to upstream xgrammar for this

**Comment by aarnphm** [comment]
iirc you can use triton kernels for using structured outputs with TPU here.

regardless I still think that having the pallas kernel is probably a more sensible path going forward here.
